common/src/main/java/io/druid/math/expr/Function.java

```diff
@@ -615,18 +684,122 @@ public String name()
+  {
+    @Override
+    public String name()
+    {
+      return "cast";
+    }
+
+    @Override
+    protected ExprEval eval(ExprEval x, ExprEval y)
+    {
+      String castTo = y.stringValue();
+      if ("string".equals(castTo)) {
+        return ExprEval.of(x.value() == null ? null : String.valueOf(x.value()));
+      }
+      if ("long".equals(castTo)) {
+        return ExprEval.of(x.type() == ExprType.STRING ? Long.valueOf(x.stringValue()) : x.longValue());
+      }
+      if ("double".equals(castTo)) {
+        return ExprEval.of(x.type() == ExprType.STRING ? Double.valueOf(x.stringValue()) : x.doubleValue());
+      }
+      throw new IllegalArgumentException("invalid type " + castTo);
+    }
+  }
+
+  class TimestampFromEpochFunc implements Function
+  {
+    // yyyy-MM-ddThh:mm:ss[.sss][Z|[+-]hh:mm]
+    private static final DateFormat ISO8601 = new ISO8601DateFormat();  // thread-safe
+
+    @Override
+    public String name()
+    {
+      return "timestamp";
+    }
+
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      if (args.isEmpty()) {
+        throw new IllegalArgumentException("function 'timestampFromEpoch' needs at least 1 argument");
+      }
+      ExprEval value = args.get(0).eval(bindings);
+      if (value.type() != ExprType.STRING) {
+        throw new IllegalArgumentException("first argument should be string type but got " + value.type() + " type");
+      }
+
+      DateFormat formatter = ISO8601;
+      if (args.size() > 1) {
+        ExprEval format = args.get(1).eval(bindings);
+        if (format.type() != ExprType.STRING) {
+          throw new IllegalArgumentException("first argument should be string type but got " + format.type() + " type");
+        }
+        formatter = new SimpleDateFormat(format.stringValue());
+      }
+      Date date;
+      try {
+        date = formatter.parse(value.stringValue());
+      }
+      catch (ParseException e) {
+        throw new IllegalArgumentException("invalid value " + value.stringValue());
+      }
+      return toValue(date);
+    }
+
+    protected ExprEval toValue(Date date)
+    {
+      return ExprEval.of(date.getTime());
+    }
+  }
+
+  class UnixTimestampFunc extends TimestampFromEpochFunc
+  {
+    @Override
+    public String name()
+    {
+      return "unix_timestamp";
+    }
+
+    @Override
+    protected final ExprEval toValue(Date date)
+    {
+      return ExprEval.of(date.getTime() / 1000);
+    }
+  }
+
+  class NvlFunc implements Function
+  {
+    @Override
+    public String name()
+    {
+      return "nvl";
+    }
+
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      if (args.size() != 2) {
+        throw new RuntimeException("function 'nvl' needs 2 argument");
+      }
+      ExprEval eval = args.get(0).eval(bindings);
+      if (eval.value() == null || (eval.type() == ExprType.STRING && Strings.isNullOrEmpty(eval.stringValue()))) {
```

Comment:
I think this is different behavior from SQL NVL which doesn't treat empty strings as null. If this behavior is required, would it make sense to make an additional function to handle the nullOrEmpty case?